pkg/overlay/overlay.go

```diff
@@ -109,15 +110,46 @@ func RemoveTemp(contentDir string) error {
 	return os.RemoveAll(contentDir)
 }
 
-// Unmount the overlay mountpoint
-func Unmount(contentDir string) (Err error) {
+func unmount(contentDir string) error {
 	mergeDir := filepath.Join(contentDir, "merge")
-	if err := unix.Unmount(mergeDir, 0); err != nil && !os.IsNotExist(err) {
+
+	if unshare.IsRootless() {
+		// Attempt to unmount the FUSE mount using either fusermount or fusermount3.
+		// If they fail, fallback to unix.Unmount
+		for _, v := range []string{"fusermount3", "fusermount"} {
+			err := exec.Command(v, "-u", mergeDir).Run()
+			if err != nil && errors.Cause(err) != exec.ErrNotFound {
+				logrus.Debugf("Error unmounting %s with %s - %v", mergeDir, v, err)
+			}
+			if err == nil {
+				return nil
+			}
+		}
+		// If fusermount|fusermount3 failed to unmount the FUSE file system, attempt unmount
+	}
+
+	// Ignore EINVAL as the specified merge dir is not a mount point
+	if err := unix.Unmount(mergeDir, 0); err != nil && err != unix.EINVAL {
 		return errors.Wrapf(err, "unmount overlay %s", mergeDir)
 	}
 	return nil
 }
 
+// Unmount the overlay mountpoint
+func Unmount(contentDir string) error {
+	files, err := ioutil.ReadDir(contentDir)
+	if err != nil {
+		return errors.Wrapf(err, "read directory")
+	}
+	for _, f := range files {
+		dir := filepath.Join(contentDir, f.Name())
+		if err := unmount(dir); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
```

Comment:
`run_linux.go` calls `RemoveTemp()` with a directory created by `TempDir()`, which in turn calls `Unmount()`, but the rewritten `Unmount()` appears to expect to be called with the parent of such a directory, as `CleanupContent()` now does, leading to multiple errors being logged at the debug level when I try this out on the command line.

Reply:
thanks.  I've restored the previous version of`Unmount()` and moved the new logic into `CleanupContent()`

Reply:
`CleanupContent()` previously attempted to compute the path that it passed to `unix.Unmount()` without a component for name of the temporary directory in it (something like "${containerdir}/overlay/merge", when we mount overlays at something that looks like "${containerdir}/overlay/${random}/merge"), so it also didn't work correctly without this patch.